File: core/src/com/xenoage/zong/core/music/beam/Beam.java
package com.xenoage.zong.core.music.beam;

import com.xenoage.utils.annotations.NonNull;
import com.xenoage.utils.math.Fraction;
import com.xenoage.zong.core.music.MusicElementType;
import com.xenoage.zong.core.music.WaypointPosition;
import com.xenoage.zong.core.music.chord.Chord;
import com.xenoage.zong.core.music.util.DurationInfo;
import com.xenoage.zong.core.position.MP;
import com.xenoage.zong.core.position.MPContainer;
import com.xenoage.zong.core.position.MPElement;
import lombok.Getter;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static com.xenoage.utils.collections.CollectionUtils.alist;
import static com.xenoage.utils.collections.CollectionUtils.getFirst;
import static com.xenoage.utils.kernel.Range.range;
import static com.xenoage.utils.math.MathUtils.min;
import static java.lang.Math.max;


/**
 * Class for a beam that connects two or more chords.
 *
 * @author Andreas Wenger
 */
public final class Beam
	implements MPElement {

	/** Spread of this beam within a system. */
	public enum VerticalSpan {
		SingleStaff,
		TwoAdjacentStaves,
		Other;
	}

	/** The waypoints in this beam. */
	@NonNull @Getter private List<BeamWaypoint> waypoints;

	//cache
	private VerticalSpan verticalSpan = null;
	private int upperStaffIndex = -1;
	private int lowerStaffIndex = -1;


	/**
	 * Creates a new beam consisting of the given waypoints.
	 * All waypoints must be in the same measure column and must be sorted by beat.
	 */
	public static Beam beam(List<BeamWaypoint> waypoints) {
		Beam beam = new Beam(waypoints);
		beam.check();
		return beam;
	}

	/**
	 * Creates a new beam consisting of the given chords with no subdivisions.
	 */
	public static Beam beamFromChords(List<Chord> chords) {
		Beam ret = beamFromChordsNoCheck(chords);
		ret.check();
		return ret;
	}

	/**
	 * Creates a new beam consisting of the given chords with no subdivisions.
	 * For testing: No parameter checks! TODO: DEPENDENCYINJECTION
	 */
	public static Beam beamFromChordsNoCheck(List<Chord> chords) { //TIDY
		List<BeamWaypoint> waypoints = alist(chords.size());
		for (Chord chord : chords) {
			waypoints.add(new BeamWaypoint(chord, false));
		}
		return new Beam(waypoints);
	}

	private Beam(List<BeamWaypoint> waypoints) {
		this.waypoints = waypoints;
	}

	/**
	 * Checks the correctness of the beam:
	 * The beam must have at least one line
	 * It must have at least 2 chords, must exist in a single measure column
	 * and the chords must be sorted by beat.
	 */
	private void check() {

		if (getMaxLinesCount() == 0)
			throw new IllegalArgumentException("A beam must have at least one line");

		if (waypoints.size() < 2)
			throw new IllegalArgumentException("At least two chords are needed to create a beam!");

		Fraction lastBeat = null;
		boolean wasLastChordGrace = false;
		int measure = getFirst(waypoints).getChord().getMP().measure;
		for (BeamWaypoint wp : waypoints) {
			MP mp = wp.getChord().getMP();

			//check, if all chords are in the same measure column
			if (mp.measure != measure)
				throw new IllegalArgumentException("A beam may only span over one measure column");

			//check, if chords are sorted by beat.
			//for grace notes, the same beat is ok.
			if (lastBeat != null) {
				int compare = mp.beat.compareTo(lastBeat);
				if ((false == wasLastChordGrace && compare <= 0) ||
						(wasLastChordGrace && compare < 0))
					throw new IllegalArgumentException("Beamed chords must be sorted by beat");
			}
			lastBeat = mp.beat;
			wasLastChordGrace = wp.getChord().isGrace();
		}
	}
	
	
	public int size() {
		return waypoints.size();
	}


	public BeamWaypoint getStart() {
		return waypoints.get(0);
	}


	public BeamWaypoint getStop() {
		return waypoints.get(waypoints.size() - 1);
	}


	/**
	 * Gets the position of the given waypoint.
	 */
	public WaypointPosition getWaypointPosition(BeamWaypoint wp) {
		return getWaypointPosition(wp.getChord());
	}


	/**
	 * Gets the position of the given chord: Start, Stop or Continue.
	 */
	public WaypointPosition getWaypointPosition(Chord chord) {
		int index = getWaypointIndex(chord);
		if (index == 0)
			return WaypointPosition.Start;
		else if (index == waypoints.size() - 1)
			return WaypointPosition.Stop;
		else
			return WaypointPosition.Continue;
	}
	
	/**
	 * Gets the index of the given chord within the beam.
	 */
	public int getWaypointIndex(Chord chord) {
		for (int i : range(waypoints))
			if (chord == waypoints.get(i).getChord())
				return i;
		throw new IllegalArgumentException("Given chord is not part of this beam.");
	}


	/**
	 * Gets the vertical spanning of this beam.
	 */
	public VerticalSpan getVerticalSpan() {
		if (verticalSpan == null)
			computeSpan();
		return verticalSpan;
	}


	/**
	 * Gets the index of the topmost staff this beam belongs to.
	 */
	public int getUpperStaffIndex() {
		if (upperStaffIndex == -1)
			computeSpan();
		return upperStaffIndex;
	}


	/**
	 * Gets the index of the bottommost staff this beam belongs to.
	 */
	public int getLowerStaffIndex() {
		if (lowerStaffIndex == -1)
			computeSpan();
		return lowerStaffIndex;
	}


	/**
	 * Replaces the given old chord with the given new one.
	 */
	public void replaceChord(Chord oldChord, Chord newChord) {
		for (int i : range(waypoints)) {
			if (waypoints.get(i).getChord() == oldChord) {
				waypoints.get(i).setChord(newChord);
				computeSpan();
				return;
			}
		}
		throw new IllegalArgumentException("Given chord is not part of this beam");
	}


	/**
	 * Returns true, if a beam lines subdivision ends at the chord
	 * with the given index.
	 */
	public boolean isEndOfSubdivision(int chordIndex) {
		return waypoints.get(chordIndex).isSubdivision();
	}


	/**
	 * Gets the chord with the given index.
	 */
	public Chord getChord(int chordIndex) {
		return waypoints.get(chordIndex).getChord();
	}


	/**
	 * Computes the vertical span of this beam.
	 */
	private void computeSpan() {
		int minStaffIndex = Integer.MAX_VALUE;
		int maxStaffIndex = Integer.MIN_VALUE;

		//check if the beam spans over a single staff or two adjacent staves or more
		Set<Integer> staves = new HashSet<Integer>();
		for (BeamWaypoint waypoint : waypoints) {
			Chord chord = waypoint.getChord();
			MP mpChord = MP.getMP(chord);
			minStaffIndex = Math.min(minStaffIndex, mpChord.staff);
			maxStaffIndex = max(maxStaffIndex, mpChord.staff);
			staves.add(mpChord.staff);
		}
		VerticalSpan verticalSpan = VerticalSpan.Other;
		if (staves.size() == 1) {
			verticalSpan = VerticalSpan.SingleStaff;
		}
		else if (staves.size() == 2) {
			Integer[] twoStaves = staves.toArray(new Integer[0]);
			if (Math.abs(twoStaves[0] - twoStaves[1]) <= 1) {
				verticalSpan = VerticalSpan.TwoAdjacentStaves;
			}
		}

		this.verticalSpan = verticalSpan;
		this.upperStaffIndex = minStaffIndex;
		this.lowerStaffIndex = maxStaffIndex;
	}
	
	/**
	 * Gets the maximum number of beam lines used in the this beam.
	 */
	public int getMaxLinesCount() {
		Fraction minDuration = null;
		for (BeamWaypoint waypoint : waypoints)
			minDuration = min(minDuration, waypoint.getChord().getDisplayedDuration());
		return DurationInfo.getFlagsCount(minDuration);
	}

	@Override public MusicElementType getMusicElementType() {
		return MusicElementType.Beam;
	}


	/**
	 * The parent of the beam is defined as the voice of the start of the beam.
	 */
	@Override public MPContainer getParent() {
		return getFirst(waypoints).getChord().getParent();
	}

	/**
	 * The MP of the beam is the same as the MP of the first chord in the beam.
	 */
	@Override public MP getMP() {
		return getFirst(waypoints).getChord().getMP();
	}

}

File: core/src/com/xenoage/zong/core/music/beam/Beam.kt
package com.xenoage.zong.core.music.beam

import com.xenoage.utils.checkIndex
import com.xenoage.utils.math.Fraction
import com.xenoage.utils.throwEx
import com.xenoage.zong.core.music.Voice
import com.xenoage.zong.core.music.WaypointPosition
import com.xenoage.zong.core.music.WaypointPosition.*
import com.xenoage.zong.core.music.beam.Beam.VerticalSpan.Other
import com.xenoage.zong.core.music.chord.Chord
import com.xenoage.zong.core.music.util.flagsCount
import com.xenoage.zong.core.position.MP
import com.xenoage.zong.core.position.MP.Companion.unknown
import com.xenoage.zong.core.position.MPElement
import kotlin.math.max
import kotlin.math.min


/**
 * Class for a beam that connects two or more chords.
 *
 * A beam can be placed within a single staff (the common case) or
 * cross two staves (e.g. in a piano score). When crossing two staves,
 * the beam belongs to the staff of the first chord.
 */
class Beam(
		/** The waypoints in this beam */
		val waypoints: List<BeamWaypoint>
) : MPElement {

	//cache
	private var _verticalSpan: VerticalSpan? = null
	private var _upperStaffIndex = -1
	private var _lowerStaffIndex = -1

	val start: BeamWaypoint
		get() = waypoints.first()

	val stop: BeamWaypoint
		get() = waypoints.last()

	/** The maximum number of beam lines used in the this beam */
	val maxLinesCount: Int
		get() {
			val minDuration = waypoints.asSequence().map{ it.chord.duration }.min() ?: throw IllegalStateException()
			return minDuration.flagsCount
		}

	/**
	 * The parent of the beam is defined as the voice of the start of the beam.
	 */
	override val parent: Voice?
		get() = waypoints.first().chord.parent

	/**
	 * The [MP] of the beam is the same as the [MP] of the first chord in the beam.
	 */
	override val mp: MP?
		get() = waypoints.first().chord.mp

	/** Spread of this beam within a system.  */
	enum class VerticalSpan {
		/** Beam within a single staff.  */
		SingleStaff,
		/** Beam crossing two adjacent staves.  */
		CrossStaff,
		/** Other span (not supported).  */
		Other
	}

	/**
	 * Checks the correctness of the beam:
	 * The beam must have at least one line
	 * It must have at least 2 chords, must exist in a single measure column
	 * and the chords must be sorted by beat.
	 */
	private fun check() {

		if (maxLinesCount == 0)
			throw IllegalArgumentException("A beam must have at least one line")

		if (waypoints.size < 2)
			throw IllegalArgumentException("At least two chords are needed to create a beam!")

		var lastBeat: Fraction? = null
		var wasLastChordGrace = false
		val startMeasure = mp?.measure
		for (wp in waypoints) {
			val chordMP = wp.chord.mp ?: continue //if unknown MP, this waypoint can not be checked
			val (_, wpMeasure, _, beat) = chordMP

			//check, if all chords are in the same measure column
			if (wpMeasure != startMeasure)
				throw IllegalArgumentException("A beam may only span over one measure column")

			//check, if chords are sorted by beat.
			//for grace notes, the same beat is ok.
			if (lastBeat != null && beat != null) {
				val compare = beat.compareTo(lastBeat)
				if (false == wasLastChordGrace && compare <= 0 || wasLastChordGrace && compare < 0)
					throw IllegalArgumentException("Beamed chords must be sorted by beat")
			}
			lastBeat = beat
			wasLastChordGrace = wp.chord.isGrace
		}
	}

	/** The number of chords in this beam. */
	val size: Int = waypoints.size

	/** Gets the position of the given waypoint. */
	fun getWaypointPosition(wp: BeamWaypoint): WaypointPosition =
		getWaypointPosition(wp.chord)

	/** Gets the position of the given chord. */
	fun getWaypointPosition(chord: Chord): WaypointPosition {
		return when (getWaypointIndex(chord)) {
			0 -> Start
			waypoints.size - 1 -> Stop
			else -> Continue
		}
	}

	/** Gets the index of the given chord within the beam. */
	fun getWaypointIndex(chord: Chord): Int =
			waypoints.indexOfFirst { it.chord === chord }.checkIndex { "Given chord is not part of this beam." }

	/** The vertical spanning of this beam. */
	val verticalSpan: VerticalSpan
			get() {
				if (_verticalSpan == null)
					computeSpan()
				return _verticalSpan!!
			}

	/** The index of the topmost staff this beam belongs to. */
	val getUpperStaffIndex: Int
		get() {
			if (_upperStaffIndex == -1)
				computeSpan()
			return _upperStaffIndex
		}


	/** The index of the bottommost staff this beam belongs to. */
	val getLowerStaffIndex: Int
		get() {
			if (_lowerStaffIndex == -1)
				computeSpan()
			return _lowerStaffIndex
		}

	/** Replaces the given old chord with the given new one. */
	fun replaceChord(oldChord: Chord, newChord: Chord) {
		for (i in waypoints.indices) {
			if (waypoints[i].chord === oldChord) {
				oldChord.beam = null
				waypoints[i].chord = newChord
				newChord.beam = this
				computeSpan()
				return
			}
		}
		throwEx("Given chord is not part of this beam")
	}

	/**
	 * Returns true, if a beam lines subdivision ends at the chord
	 * with the given index.
	 */
	fun isEndOfSubdivision(chordIndex: Int): Boolean =
		waypoints[chordIndex].subdivision

	/** Gets the chord with the given index. */
	fun getChord(chordIndex: Int): Chord =
		waypoints[chordIndex].chord

	/** Computes the vertical span of this beam. */
	private fun computeSpan() {
		var minStaffIndex = Integer.MAX_VALUE
		var maxStaffIndex = Integer.MIN_VALUE

		//check if the beam spans over a single staff or two adjacent staves or more
		for (waypoint in waypoints) {
			val chord = waypoint.chord
			val mpChord = chord.mp
			if (mpChord == null || mpChord.staff == unknown) { //unknown MP? then we can not compute this beam
				_verticalSpan = Other
				_upperStaffIndex = unknown
				_lowerStaffIndex = unknown
				return
			}
			minStaffIndex = min(minStaffIndex, mpChord.staff)
			maxStaffIndex = max(maxStaffIndex, mpChord.staff)
		}
		var verticalSpan = Other
		if (maxStaffIndex == minStaffIndex)
			verticalSpan = VerticalSpan.SingleStaff
		else if (maxStaffIndex - minStaffIndex == 1)
			verticalSpan = VerticalSpan.CrossStaff

		_verticalSpan = verticalSpan
		_upperStaffIndex = minStaffIndex
		_lowerStaffIndex = maxStaffIndex
	}

	companion object {

		/**
		 * Creates a new beam consisting of the given chords with no subdivisions.
		 * OBSOLETE
		 */
		fun beamFromChords(chords: List<Chord>): Beam {
			val ret = beamFromChordsUnchecked(chords)
			ret.check()
			return ret
		}

		/**
		 * Creates a new beam consisting of the given chords with no subdivisions.
		 * For testing: No parameter checks!
		 * OBSOLETE
		 */
		fun beamFromChordsUnchecked(chords: List<Chord>): Beam =
				Beam(chords.map { BeamWaypoint(it) })

	}

}
